src/Chapter4.hs

```diff
@@ -472,10 +478,12 @@ Implement the Applicative instance for our 'Secret' data type from before.
 -}
 instance Applicative (Secret e) where
     pure :: a -> Secret e a
-    pure = error "pure Secret: Not implemented!"
+    pure x = Reward x
 
     (<*>) :: Secret e (a -> b) -> Secret e a -> Secret e b
-    (<*>) = error "(<*>) Secret: Not implemented!"
+    (<*>) (Reward func) (Reward r) = Reward (func r)
+    (<*>) (Trap t) _ = Trap t
+    (<*>) _ (Trap t) = Trap t
```

Comment:
This is correct!
But you can use the `Functor` instance of `Secret` to implement this instead of the 1st and 3rd cases 🙂

Reply:
Great tip! Refactored this to reuse fmap 🙂